src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -0,0 +1,583 @@
+                            return response;
+                        }
+
+                        // Windows auth
+                        if (STSAuthHelper.TryRetrieveSTSToken(_baseUri, CredentialStore.Instance, response))
+                        {
+                            // Auth token found, create a new message handler and retry.
+                            await UpdateHttpClient();
+                            continue;
+                        }
+
+                        // Prompt the user
+                        promptCredentials = await PromptForCredentials(cancellationToken);
+
+                        if (promptCredentials != null)
+                        {
+                            // The user entered credentials, create a new message handler that includes
+                            // these and retry.
+                            await UpdateHttpClient(promptCredentials);
+                            continue;
+                        }
+                    }
+                    finally
+                    {
+                        _httpClientLock.Release();
+                    }
+                }
+
+                if (promptCredentials != null && HttpHandlerResourceV3.CredentialsSuccessfullyUsed != null)
+                {
+                    HttpHandlerResourceV3.CredentialsSuccessfullyUsed(_baseUri, promptCredentials);
+                }
+
+                return response;
+            }
+        }
+
+        private async Task<ICredentials> PromptForCredentials(CancellationToken cancellationToken)
+        {
+            ICredentials promptCredentials = null;
+
+            if (HttpHandlerResourceV3.PromptForCredentials != null)
+            {
+                try
+                {
+                    // Only one prompt may display at a time.
+                    await _credentialPromptLock.WaitAsync();
+
+                    promptCredentials =
+                        await HttpHandlerResourceV3.PromptForCredentials(_baseUri, cancellationToken);
+                }
+                finally
+                {
+                    _credentialPromptLock.Release();
+                }
+            }
+
+            return promptCredentials;
+        }
+
+        private async Task UpdateHttpClient()
+        {
+            // Get package source credentials
+            var credentials = CredentialStore.Instance.GetCredentials(_baseUri);
+
+            if (credentials == null
+                && !String.IsNullOrEmpty(_packageSource.UserName)
+                && !String.IsNullOrEmpty(_packageSource.Password))
+            {
+                credentials = new NetworkCredential(_packageSource.UserName, _packageSource.Password);
+            }
+
+            if (credentials != null)
+            {
+                CredentialStore.Instance.Add(_baseUri, credentials);
+            }
+
+            await UpdateHttpClient(credentials);
+        }
+
+        private async Task UpdateHttpClient(ICredentials credentials)
```

Comment:
I thought we discussed that you found a way to get rid of replacing the HttpClient